File: PythonSpark/JobManagment/JobRunner.py
import logging
from typing import Type

from pyspark.sql import SparkSession

from JobManagment.Job import Job
from JobManagment.Runner import Runner


class JobRunner(Runner):
    __registered__ = {}

    def register(self, job_name: str, job_class: Type[Job]):
        logging.info(f"Stared registering {job_name} job")
        self.__registered__[job_name] = job_class
        logging.info(f"Finished registering {job_name} job")

    def run(self, job_name: str, spark_master: str, spark_app_name: str, args: list):
        logging.info(f"Stared running {job_name} job on a spark with master {spark_master}")
        job_class = self.__registered__[job_name]

        arguments = job_class.parser().parse_args(args)

        spark = SparkSession.builder.master(spark_master).appName(spark_app_name).getOrCreate()

        job = job_class(arguments, spark)
        logging.info("Started loading data")
        data = job.load()
        logging.info("Finished loading data")

        logging.info("Started processing data")
        result = job.process(data)
        logging.info("Finished processing data")

        logging.info("Started writing results")
        job.write(result)
        logging.info("Finished writing results")

        logging.info(f"Finished running {job_name} job on a spark with master {spark_master}")

File: ScalaSpark/src/main/scala/JobManagment/JobRunner.scala
package JobManagment

import Utils.ArgumentsParsing.parseArguments
import de.halcony.argparse.Parser
import org.apache.log4j.{Level, Logger}
import org.apache.spark.sql.SparkSession

class JobRunner extends Runner {
  private var registeredJobs: Map[String, (Class[Job], Parser)] = Map()

  override def register[T <: Job](jobName: String, jobClass: Class[T], argumentsParser: Parser): Unit = {
    registeredJobs = registeredJobs + (jobName -> (jobClass.asInstanceOf[Class[Job]], argumentsParser))
  }

  override def run(jobName: String, sparkMaster: String, sparkAppName: String, args: Array[String]): Unit = {
    val jobTuple = registeredJobs(jobName)

    val parser = jobTuple._2

    val arguments = parseArguments(args, parser)

    Logger.getLogger("org").setLevel(Level.ERROR)

    val spark = SparkSession.builder().master(sparkMaster).appName(sparkAppName).getOrCreate()

    val job = getJobInstance(arguments, spark, jobTuple._1)

    val data = job.load()

    val result = job.process(data)

    job.write(result)
  }

  private def getJobInstance[T](arguments: Map[String, String], spark: SparkSession, clazz: Class[T]): T = {
    val constructors = clazz.getConstructors

    val constructor = constructors(0)

    constructor.newInstance(arguments, spark).asInstanceOf[T]
  }
}
